### ISSUE TITLE
Caffeine - Temporary Keep-Awake

### ISSUE THREAD

**Issue by LukeAbellSmith** [GrapheneOS/os-issue-tracker#1007]
A drawer widget would be ideal.
It is possible to achieve a similar result by going into settings and changing the sleep timeout if the user is in a safe location where having the device unlocked poses no risk, but then the user needs to remember to set the timeout shorter again when leaving so as to not compromise security by forgetting and leaving the device unlocked and unattended.

**Comment by ph00lt0** [comment]
You could probably just just app-pinning for this. That will also limit the risk that someone opens different apps if left unattended. @LukeAbellSmith. Leaving the device fully open never really is a good idea.

https://support.google.com/android/answer/9455138?hl=en

**Comment by LukeAbellSmith** [comment]
It's pretty safe if I'm home alone.
Also if you're arguing that, the maximum screen lock timeout should probably be reduced from 30 minutes.

**Comment by chwallen** [comment]
> You could probably just just app-pinning for this.

App pinning does not prevent the screen lock timeout, so it is unfortunately not a work-around.

> Leaving the device fully open never really is a good idea.

If I am in my kitchen following a recipe and I want the screen to remain active for 5 or 10 minutes (or a similar scenario), this is really useful. It is annoying having to either repeatedly tap my device to keep the screen active, or temporarily change the screen timeout period to 10 minutes and always forgetting to turn to back down.

**Comment by LukeAbellSmith** [comment]
This is exactly what I'm talking about. :)

**Comment by lazytownfan** [comment]
This is the one Quick Tile setting that I miss from LineageOS (which can toggle between 5 minutes, 10 minutes, 30 minutes, and ∞).

Currently, I have to use the [Coffee](https://f-droid.org/en/packages/com.github.muellerma.coffee/) app on GrapheneOS.  It's the only FOSS app that's currently maintained (that I'm aware of) and it works acceptably, though the Quick Tile only toggles between being turned on forever or it requires you to turn it off manually.  It's also not as fast as Caffeine on LineageOS, though I suspect this behavior happens since Coffee is just a normal app and isn't a system-wide app.

Sometimes I need to prevent my device from sleeping for a few minutes in a scenario where I have low security needs, such as if I need to quickly reference contact info to a computer or when I'm drawing from a reference picture.  I'm still present and aware of my device - I just don't want to rush to prevent my phone from sleeping every 2 minutes (or whatever people have the sleep time to).

**Comment by lokesh-krishna** [comment]
Could this work by offering a quick settings tile for the screen timeout itself? So that one doesn't need to dig into settings for it and resetting it would just mean setting it back to whatever value one usually uses.

**Comment by russianbot91** [comment]
I've found this this app works almost as well as the one included with LineageOS.

https://apt.izzysoft.de/fdroid/index/apk/tr.edu.iyte.caffeine

There is one in the official F-Droid repositories that's only 11kB, but it doesn't work on modern versions of Android.

**Comment by sevmonster** [comment]
I needed to keep the screen on while driving (reading navigation) and was sad to see that Caffiene, which has been in every custom ROM I've used in years, wasn't present the only time I truly needed it. I ended up adjusting the screen timeout from Settings at a stoplight.

> https://apt.izzysoft.de/fdroid/index/apk/tr.edu.iyte.caffeine

This didn't work for me on Pixel 6 Pro with most recent Graphene.

Coffee as mentioned above has also added a "Next timeout" quick tile, which has the time increments like Caffiene. It doesn't go to infinite as a final step though so it's still not a drop-in replacement.

**Comment by r3g-5z** [comment]
> I needed to keep the screen on while driving (reading navigation)

Whatever app you're using needs to be written in a way to keep a wakelock for this like normal navigation apps (e.g. Google Maps).

If the user has to keep the screen on for driving navigation, this is unsafe and the developer needs to correct this. This isn't hard for them to implement.

**Comment by sevmonster** [comment]
> Whatever app you're using needs to be written in a way to keep a wakelock for this like normal navigation apps (e.g. Google Maps).

It was a text document, not navigation app. Point me to a text reader that supports wake lock and I'll gladly try it.

**Comment by lokesh-krishna** [comment]
[ReadEra](https://play.google.com/store/apps/details?id=org.readera) does

**Comment by sevmonster** [comment]
> I needed to keep the screen on while driving (reading navigation) and was sad to see that Caffiene, which has been in every custom ROM I've used in years, wasn't present the only time I truly needed it. I ended up adjusting the screen timeout from Settings at a stoplight.
> 
> > https://apt.izzysoft.de/fdroid/index/apk/tr.edu.iyte.caffeine
> 
> This didn't work for me on Pixel 6 Pro with most recent Graphene.
> 
> Coffee as mentioned above has also added a "Next timeout" quick tile, which has the time increments like Caffiene. It doesn't go to infinite as a final step though so it's still not a drop-in replacement. Though, maybe it's good that it doesn't.

**Comment by soupslurpr** [comment]
`@hybridstaticanimate:matrix.org` would like to say this (they don't have a GitHub account yet):

"I feel this issue has become more important with the removal of the AOSP functionality to automatically extend the lock timer. I must stress that I previously used the AOSP feature for certain usecases and now those usecases have become far more difficult.

I propose adding a feature to extend the lock timer temporarily, via a manual button such as a power menu button and/or quicktile. This can loosely be called a security feature, since it would take a vague and hidden AOSP feature and replace it with a more robust implementation that allows it to be manual and explicit. There would be many security benefits of this feature, such as keeping the benefits of consistent lock behaviour for those who do not use it, a user needs to explicitly opt into using it, the ability to see when it is in use and the ability to manually cancel it as well (in a notification, probably), and not needing to extend the global timer to accommodate the need for this feature, allowing more users to tolerate much shorter a lock timer as a default.

This would also pair well with the planned movement lock feature, in case the device were to be swiped while the extension is active. (https://github.com/GrapheneOS/os-issue-tracker/issues/207)"

**Comment by woiwoiwoiwoi** [comment]
Yes, there are. Quite a few even. None of them, however work as well as the native integration in LineageOS or CalyxOS. There's always a delay when activating them via quick tile.

**Comment by thestinger** [comment]
@woiwoiwoiwoi Not true, multiple of those apps work fine.

**Comment by woiwoiwoiwoi** [comment]
@thestinger Well, could you tell me which one works fine for you? I tried the ones mentioned above and all maintained ones listed on F-Droid (but installed via Obtainium). Battery Optimisation is unrestricted, all permissions granted, but I still get a delay of around 5 seconds for any kind of reaction after the first tap on the tile. Following taps work fine. When the tile is then unused for a while, I again get that delay.
